rust/src/mask/masking.rs

```diff
@@ -0,0 +1,516 @@
+    ModelMismatch,
+}
+
+impl AggregatedModel {
+    pub fn new(config: MaskConfig) -> Self {
+        Self {
+            nb_models: 0,
+            model: MaskObject {
+                data: vec![],
+                config: config.clone(),
+            },
+            config,
+        }
+    }
+
+    pub fn validate_unmasking(&self, mask: &Mask) -> Result<(), UnmaskingError> {
+        // We cannot perform unmasking without at least one real model
+        if self.nb_models == 0 {
+            return Err(UnmaskingError::NoModel);
+        }
+
+        if self.nb_models > self.config.model_type.nb_models_max() {
+            return Err(UnmaskingError::TooManyModels);
+        }
+
+        if self.config != mask.config || self.model.data.len() != mask.data.len() {
+            return Err(UnmaskingError::MaskMismatch);
+        }
+
+        if !mask.is_valid() {
+            return Err(UnmaskingError::InvalidMask);
+        }
+
+        Ok(())
+    }
+
+    pub fn unmask(mut self, mask: Mask) -> Model {
+        let scaled_add_shift = self.config.add_shift() * BigInt::from(self.nb_models);
+        let exp_shift = self.config.exp_shift();
+        let order = self.config.order();
+        self.model
+            .data
+            .drain(..)
+            .into_iter()
+            .zip(mask.data.into_iter())
+            .map(|(masked_weight, mask)| {
+                // PANIC_SAFE: The substraction panics if it
+                // underflows, which can only happen if:
+                //
+                //     mask > self.config.order()
+                //
+                // If the mask is valid, we are guaranteed that this
+                // cannot happen. Thus this method may panic only if
+                // given an invalid mask.
+                let n = (masked_weight + order.clone() - mask) % order.clone();
+
+                // UNWRAP_SAFE: to_bigint never fails for BigUint
+                let ratio = Ratio::<BigInt>::from(n.to_bigint().unwrap());
+
+                ratio / exp_shift.clone() - &scaled_add_shift
+            })
+            .collect()
+    }
+
+    pub fn validate_aggregation(&self, model: &MaskedModel) -> Result<(), AggregationError> {
+        if self.config != model.config || self.model.data.len() != model.data.len() {
+            return Err(AggregationError::ModelMismatch);
+        }
+
+        if self.nb_models == self.config.model_type.nb_models_max() {
+            return Err(AggregationError::TooManyModels);
+        }
+
+        if !model.is_valid() {
+            return Err(AggregationError::InvalidModel);
+        }
+
+        Ok(())
+    }
+
+    pub fn aggregate(&mut self, model: MaskedModel) {
+        if self.nb_models == 0 {
+            self.model = model;
+            self.nb_models = 1;
+            return;
+        }
+
+        let order = self.config.order();
+        for (i, j) in self.model.data.iter_mut().zip(model.data.into_iter()) {
+            *i = (i.clone() + j) % order.clone()
```

Comment:
So this works:

```rust
        for (i, j) in self.object.data.iter_mut().zip(object.data.into_iter()) {
            *i = (&*i + j) % &order
        }
```

Somehow `rustc` doesn't like the fact that `i` is `&mut BigUint` instead of just `&BigUint`.

Reply:
might be because `num` only implements the `Add` trait for `BigUint` and `&BigUint` but not for `&mut BigUint`. but it's nice that deref and ref again is a feasible workaround.